src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -802,122 +836,92 @@ namespace QgsWms
+            {
+              layerName = layerAliasIt.value();
+            }
+
+            layerElement.setAttribute( QStringLiteral( "name" ), layerName );
             getFeatureInfoElement.appendChild( layerElement );
+            if ( sia2045 ) //the name might not be unique after alias replacement
+            {
+              layerElement.setAttribute( QStringLiteral( "id" ), layer->id() );
+            }
           }
 
-          QgsRasterLayer *rasterLayer = qobject_cast<QgsRasterLayer *>( currentLayer );
-          if ( rasterLayer )
+          if ( layer->type() == QgsMapLayer::VectorLayer )
           {
-            if ( !infoPoint )
+            QgsVectorLayer *vectorLayer = qobject_cast<QgsVectorLayer *>( layer );
+            if ( vectorLayer )
             {
-              continue;
-            }
-            QgsPointXY layerInfoPoint = mapSettings.mapToLayerCoordinates( currentLayer, *( infoPoint.get() ) );
-            if ( !featureInfoFromRasterLayer( rasterLayer, mapSettings, &layerInfoPoint, result, layerElement, version, infoFormat ) )
-            {
-              continue;
+              if ( !featureInfoFromVectorLayer( vectorLayer, infoPoint.get(), featureCount, result, layerElement, mapSettings, renderContext, version, mWmsParameters.infoFormatAsString(), featuresRect.get(), filterGeom.get() ) )
+              {
+                break;
+              }
+              break;
             }
           }
           else
           {
-            continue;
+            if ( infoFormat == QgsWmsParameters::Format::GML )
+            {
+              layerElement = result.createElement( QStringLiteral( "gml:featureMember" )/*wfs:FeatureMember*/ );
+              getFeatureInfoElement.appendChild( layerElement );
+            }
+
+            QgsRasterLayer *rasterLayer = qobject_cast<QgsRasterLayer *>( layer );
+            if ( rasterLayer )
+            {
+              if ( !infoPoint )
+              {
+                break;
+              }
+              QgsPointXY layerInfoPoint = mapSettings.mapToLayerCoordinates( layer, *( infoPoint.get() ) );
+              if ( !featureInfoFromRasterLayer( rasterLayer, mapSettings, &layerInfoPoint, result, layerElement, version, mWmsParameters.infoFormatAsString() ) )
+              {
+                break;
+              }
+              break;
+            }
           }
+          break;
         }
       }
     }
 
     if ( featuresRect )
     {
-      if ( infoFormat.startsWith( QLatin1String( "application/vnd.ogc.gml" ) ) )
+      if ( infoFormat == QgsWmsParameters::Format::GML )
       {
         QDomElement bBoxElem = result.createElement( QStringLiteral( "gml:boundedBy" ) );
         QDomElement boxElem;
-        int gmlVersion = infoFormat.startsWith( QLatin1String( "application/vnd.ogc.gml/3" ) ) ? 3 : 2;
+        int gmlVersion = mWmsParameters.infoFormatAsString().startsWith( QLatin1String( "application/vnd.ogc.gml/3" ) ) ? 3 : 2;
```

Comment:
I'd suggest to add a *QgsWmsParameters::infoFormatVersion()* to manage the GML version as it's done several times.